File: 백준/Gold/14502. 연구소/연구소.cc
#include<iostream>
#include<algorithm>
#include<string>
#include<queue>
#include<unordered_map>
using namespace std;

int n, m;
int a[9][9];
int tmp[9][9];
vector<pair<int, int>> blank, virus;
int dy[] = { -1, 0, 1, 0 };
int dx[] = { 0, 1, 0, -1 };
int ret = 0;

void dfs(int y, int x) {
	tmp[y][x] = 2;
	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
		if (tmp[ny][nx] != 0) continue;
		dfs(ny, nx);
	}
}

void solve(int now, int cnt) {
	if (cnt == 3) {
		int sum = 0;
		copy(&a[0][0], &a[0][0] + 9 * 9, &tmp[0][0]);
		for (pair<int, int> p : virus) {
			int y = p.first;
			int x = p.second;
			dfs(y, x);
		}
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (tmp[i][j] == 0) {
					sum++;
				}
			}
		}
		ret = max(ret, sum);
		return;
	}
	for (int i = now; i < blank.size(); i++) {
		int y = blank[i].first;
		int x = blank[i].second;
		a[y][x] = 1;
		solve(i + 1, cnt + 1);
		a[y][x] = 0;
	}
}

int main() {
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> a[i][j];
			if (a[i][j] == 0) {
				blank.push_back({ i, j });
			}
			else if (a[i][j] == 2) {
				virus.push_back({ i, j });
			}
		}
	}
	solve(0, 0);
	cout << ret;
}

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int n, m;
    static int[][] a;
    static ArrayList<Pair> list = new ArrayList<>();
    static ArrayList<Pair> virus = new ArrayList<>();
    static int[] dy = { -1, 0, 1, 0};
    static int[] dx = {0, 1, 0, -1};
    static int ret = 0;
    
    static class Pair{
        int y, x;
        Pair(int y, int x){
            this.y = y;
            this.x = x;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        a = new int[n][m];
        for(int i = 0; i < n; i++){
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j < m; j++){
                a[i][j] = Integer.parseInt(st.nextToken());
                list.add(new Pair(i, j));
                if(a[i][j] == 2){
                    virus.add(new Pair(i, j));
                }
            }
        }
        solve(0, 0);
        System.out.println(ret);
    }

    private static void solve(int cnt, int now){
        if(cnt == 3){
            bfs();
            return;
        }
        for(int i = now; i < list.size(); i++){
            Pair cur = list.get(i);
            if(a[cur.y][cur.x] != 0) continue;
            a[cur.y][cur.x] = 1;
            solve(cnt + 1, now + 1);
            a[cur.y][cur.x] = 0;
        }
    }

    private static void bfs(){
        int[][] b = new int[n][m];
        for(int i = 0; i < n; i++){
            b[i] = a[i].clone();
        }
        Queue<Pair> q = new ArrayDeque<>();
        for(int i = 0; i < virus.size(); i++){
            q.offer(virus.get(i));
        }
        while(!q.isEmpty()){
            Pair cur = q.poll();
            int y = cur.y;
            int x = cur.x;
            for(int i = 0; i < 4; i++){
                int ny = y + dy[i];
                int nx = x + dx[i];
                if(ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
                if(b[ny][nx] != 0) continue;
                b[ny][nx] = 2;
                q.offer(new Pair(ny, nx));
            }
        }
        int sum = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(b[i][j] == 0){
                    sum++;
                }
            }
        }
        ret = Math.max(sum, ret);
    }
}
